processing/src/main/java/org/apache/druid/query/rowsandcols/concrete/ColumnHolderRACColumn.java

```diff
@@ -91,7 +91,7 @@ public int numRows()
       public boolean isNull(int rowNum)
       {
         offset.set(rowNum);
-        return valueSelector.isNull();
+        return valueSelector.getObject() == null;
```

Comment:
does this have a different contract than `BaseNullableColumnValueSelector`? Its `isNull` method means that `getLong`/`getFloat`/`getDouble` will return a null value, not necessarily that `getObject` is null. E.g. sometimes non-numeric selectors will implement these primitive getters, but not necessarily be null rows. 

Looking at some of the things that implement it, it doesn't appear to be different, e.g. `ObjectColumnAccessorBase` has an implementation of `getLong` that checks if the instance is a `String`, and if so, tries to parse it. This means that `getObject` will not return null, but `getLong` isn't valid, so there is no way to determine that the `0` it returns is actually a `null` because we couldn't parse the string `'hello'` into a long.

Reply:
The contract here is supposed to be abandoning the "numerical null" thing that the ValueSelectors follow.  It is explicitly "is it null" rather than "will it be null if cast to a primitive type"

Reply:
how do you know if it is safe to call `getLong` if that is the contract?

Reply:
Or maybe we should eliminate implicit casting in the numeric getter methods to make it so that they only work if the column type really is a number?

Reply:
i think we need to revisit this contract, since it doesn't really feel compatible to coexist with implicit casting on the `getLong`/`getFloat` etc methods. If you have a string column but want to read numbers from it but not all of them can be parsed into numbers, you can't reliably use any of the primitive getters, because `isNull` would say false, but `getLong` would not be parseable into a number and return 0. See `ObjectColumnAccessorBase`.

This means you can't actually use these methods if you want to try to get numbers from an object column, so they should probably instead throw exceptions instead of silently giving you zeros that should have been handled as nulls.

We can fix this by just forbidding this sort of implicit casting, I'm not sure it does us any favors in terms of making code clear to follow, while having explicit casting wrapper gizmos would make it obvious when something is turning a string into a number.

Reply:
not that it has to be done now, but it does seem problematic and why `ColumnValueSelector.isNull` has the strange contract it does, because you kind of need it if you allow implicit casting.

Reply:
The semantics as defined here is that a non-parseable String, when read from `getLong` is `0` instead of `null`.  If something wants that to be `null`, it needs to do its own parsing/handling.  It is entirely possible to, if you have a String object, use the `getObject()` form and handle it yourself.

We could make this logic be generically accessible by having a `LongAccessor` interface and ask the column to become one of those when it wants to force-cast to `Long`.  What I'm trying to say is that I understand the semantics you are looking for and I think that it's totally valid, I don't think it has to be delivered from this set of `getters` which just exist to provide primitive values in the simplest possible way.  That said, perhaps if we had the type-specific accessors we would end up not needing these generic `get` methods and then life would just be better.  Either way, it's something that can be adjusted in another PR.